### PR TITLE
Create POST API for Orchestrator Zone API

### PR DESCRIPTION
**Pull request by babamishra9**
- Updated endpoint path to /orchestrator/zones
- Added post method in controller
- Added steps and logics as per service-broker provisionZone flow.
- Added required gradle dependencies.
- Added required env variables to uaa.yaml

### CODE REVIEW

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java [hunk 1]
```diff
@@ -29,6 +113,268 @@ public ZoneResponse getZoneDetails(String zoneName) {
         return new ZoneResponse(zoneName, zone, connectionDetails);
     }
 
+    public void createZone(ZoneRequest zoneRequest) throws ZoneServiceException {
```
Comment (irinaepshteyn) [inline zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java:116]: This function needs refactoring as it's very difficult to read.  I think most of the code could be moved into its own functions with error handling.  Sort of like:
`public void createZone(ZoneRequest zoneRequest) throws ZoneServiceException {`
`    createSigningKey();`
`    setTokenPolicy();`
`    createDefaultIdp();`
`    createUserGroups();`
`...`
`}`

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java [hunk 2]
```diff
@@ -1,23 +1,107 @@
...
 
-    public ZoneService(IdentityZoneProvisioning zoneProvisioning, String uaaDashboardUri) {
+    public ZoneService(IdentityZoneProvisioning zoneProvisioning,
+                       IdentityProviderProvisioning idpProvisioning,
+                       ScimGroupProvisioning groupProvisioning,
+                       MultitenantClientServices multitenantClientServices,
+                       QueryableResourceManager<ClientDetails> clientDetailsService,
+                       ClientDetailsValidator clientDetailsValidator,
+                       String uaaDashboardUri, String clientId, String zoneAuthorities, String grantTypes,
+                       String resourceIds, String scopes, String uaaClientID, String uaaUrl
```
Comment (irinaepshteyn) [inline zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java:89]: Number of constructor parameters looks excessive.  How come neither service broker [UaaZoneProvisioner](https://github.build.ge.com/predix/security-service-broker/blob/develop/service-broker-service/src/main/java/com/ge/predix/servicebroker/service/UaaZoneProvisioner.java#L73) nor UAA [JdbcIdentityZoneProvisioning](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/server/src/main/java/org/cloudfoundry/identity/uaa/zone/JdbcIdentityZoneProvisioning.java#L49) required that many parameters?

**Review by irinaepshteyn — CHANGES_REQUESTED** [review]
Hi @babamishra9 , I submitted the initial comments.  PR needs further review; it's not quite ready for merging.

Also, considerations for the future:

Controller for native zone creation is called [IdentityZoneEndpoints](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/server/src/main/java/org/cloudfoundry/identity/uaa/zone/IdentityZoneEndpoints.java) and all related classes have IdentityZone prefix.  For example [IdentityZone](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/model/src/main/java/org/cloudfoundry/identity/uaa/zone/IdentityZone.java), [IdentityZoneConfiguration](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/model/src/main/java/org/cloudfoundry/identity/uaa/zone/IdentityZoneConfiguration.java)

It would be nice to use orchestrator prefix to distinguish orchestrator zone provisioning from native (identity) zone provisioning.  For example, OrchestratorZoneController, OrchestratorZone, OrchestratorZoneRequest, etc.

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/Zone.java [hunk 3]
```diff
@@ -5,12 +5,33 @@
...
     private String subdomain;
+
+    public String getAdminSecret() {
+        return adminSecret;
+    }
+
+    public void setAdminSecret(String adminSecret) {
+        this.adminSecret = adminSecret;
+    }
+
```
Comment (Yogeshkumarb57) [inline zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/Zone.java:29]: @Data annotation will generate getter, setter methods for adminSecret and subdomain fields so please revert this change.

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java [hunk 4]
```diff
@@ -29,6 +113,268 @@ public ZoneResponse getZoneDetails(String zoneName) {
...
+                logger.error("Unable to create client for subdomain : " + subdomain + " Zone Id : "+zoneId+" Exception is :" + e.getMessage());
+                throw new ZoneServiceException("Unexpected exception while create client for " +
+                                               "subdomain : " + subdomain + " Zone Id : "+zoneId+" Exception is :" + e.getMessage());
+            } finally{
+                updateDefaultAdminAuthorities(emptyAuthorityList, tempAuthorityList, zoneId);
+            }
+        }
+    }
+
+    public void updateDefaultAdminAuthorities(final List<SimpleGrantedAuthority> addAuthorities,
```
Comment (irinaepshteyn) [inline zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java:199]: This is done to elevate authorities of the base zone admin client so it could be used to administer newly created zone, like add an admin client.  Since here you are calling clientDetailsService.create() directly without using REST API and without token, I suspect it's not needed.

**Comment by babamishra9** [comment]
> Hi @babamishra9 , I submitted the initial comments. PR needs further review; it's not quite ready for merging.
> 
> Also, considerations for the future:
> 
> Controller for native zone creation is called [IdentityZoneEndpoints](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/server/src/main/java/org/cloudfoundry/identity/uaa/zone/IdentityZoneEndpoints.java) and all related classes have IdentityZone prefix. For example [IdentityZone](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/model/src/main/java/org/cloudfoundry/identity/uaa/zone/IdentityZone.java), [IdentityZoneConfiguration](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/model/src/main/java/org/cloudfoundry/identity/uaa/zone/IdentityZoneConfiguration.java)
> 
> It would be nice to use orchestrator prefix to distinguish orchestrator zone provisioning from native (identity) zone provisioning. For example, OrchestratorZoneController, OrchestratorZone, OrchestratorZoneRequest, etc.

Renamed classes. Done. Thanks!

File: uaa/src/main/resources/uaa.yml [hunk 5]
```diff
@@ -502,6 +502,12 @@ login:
 uaa:
   # The hostname of the UAA that this login server will connect to
   url: http://localhost:8080/uaa
+  clientId: admin
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply uaa/src/main/resources/uaa.yml:505]: To run UAA locally, I always refer and updated db configuration to [uaa.yml](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/uaa/src/main/resources/uaa.yml).
I have also put env variables to suggested [uaa.yml](https://github.com/GESoftware-CF/uaa/blob/predix_extensions_75.18.1/scripts/cargo/uaa.yml).

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/Zone.java [hunk 6]
```diff
@@ -5,12 +5,33 @@
...
 @Data
 @AllArgsConstructor
 @NoArgsConstructor
 @JsonInclude(Include.NON_NULL)
 public class Zone {
+
+    @NotNull(message = org.cloudfoundry.identity.uaa.zone.ZoneController.MANDATORY_VALIDATION_MESSAGE)
     private String adminSecret;
+
+    @NotNull(message = org.cloudfoundry.identity.uaa.zone.ZoneController.MANDATORY_VALIDATION_MESSAGE)
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/Zone.java]: Done. Thanks!

File: zone-service/build.gradle [hunk 7]
```diff
@@ -16,12 +16,25 @@ dependencies {
...
+    }
+    implementation 'org.apache.commons:commons-lang3:3.0'
     testImplementation(libraries.springBootStarterTest)
 }
 
 test {
     useJUnitPlatform()
+}
+
+tasks.withType(AbstractCompile) {
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/build.gradle:38]: Hence, sun.security.x509 packages in JDK 11, added to gradle compiler args :  options.compilerArgs += ["--add-exports", "java.base/sun.security.provider=ALL-UNNAMED"]

[Access `sun.security.x509` in JDK 11 without modules?](https://stackoverflow.com/questions/61149455/access-sun-security-x509-in-jdk-11-without-modules).

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/Zone.java [hunk 8]
```diff
@@ -5,12 +5,33 @@
...
 @Data
 @AllArgsConstructor
 @NoArgsConstructor
 @JsonInclude(Include.NON_NULL)
 public class Zone {
+
+    @NotBlank(message = org.cloudfoundry.identity.uaa.zone.ZoneController.MANDATORY_VALIDATION_MESSAGE)
     private String adminSecret;
+
+    @NotBlank(message = org.cloudfoundry.identity.uaa.zone.ZoneController.MANDATORY_VALIDATION_MESSAGE)
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/Zone.java]: subdomain is optional now. Done. Thanks!

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java [hunk 9]
```diff
@@ -29,6 +113,268 @@ public ZoneResponse getZoneDetails(String zoneName) {
...
+            zoneSigningKey = createSigningKey();
+        } catch (Exception e) {
+            logger.error("Unable to create signingKey for subdomain: " + subdomain + " Exception is : " + e.getMessage());
+            throw new ZoneServiceException("Unexpected exception while " +
+                                           "create signingKey for subdomain: ." + subdomain + " Exception is : " + e.getMessage());
+        }
+        IdentityZone identityZone = new IdentityZone();
+        identityZone.setId(id);
+        identityZone.setName(name);
+        identityZone.setSubdomain(subdomain);
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java]: Done. Thanks!

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java [hunk 10]
```diff
@@ -29,6 +113,268 @@ public ZoneResponse getZoneDetails(String zoneName) {
...
+        IdentityZone created = null;
+        try {
+            created = zoneProvisioning.create(identityZone);
+        } catch (Exception e) {
+            logger.error("Unable to create identity zone for subdomain : " + subdomain + " Exception is : " + e.getMessage());
+            throw new ZoneServiceException("Unexpected exception while create identity zone for " +
+                                           "subdomain : ." + subdomain + " Exception is : " + e.getMessage());
+        }
+
+        if(created != null){
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java]: Done.

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneController.java [hunk 11]
```diff
@@ -41,6 +45,12 @@ public ResponseEntity<ZoneResponse> getZone(@NotBlank(message = MANDATORY_VALIDA
         return new ResponseEntity<>(zoneService.getZoneDetails(name), HttpStatus.ACCEPTED);
     }
 
+    @PostMapping
+    public ResponseEntity<?> createZone(@RequestBody @Valid ZoneRequest zoneRequest ) throws ZoneServiceException {
```
[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneController.java]: Done

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java (hunk 1 above)
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java]: Done

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/Zone.java (hunk 3 above)
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/Zone.java]: Done, Thanks!

File: zone-service/build.gradle (hunk 7 above)
[the remark this answers is not in the document]
Comment (irinaepshteyn) [reply zone-service/build.gradle:38]: The link you are referring to also mentions that this is just a workaround and "is liable to stop working in the future. The solution would be to find a way to avoid depending on those classes."

You might want to explore [java.security.cert](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/cert/package-summary.html) package or [bouncy-castle](https://stackoverflow.com/questions/29852290/self-signed-x509-certificate-with-bouncy-castle-in-java)

We could address it as a separate story.

[the remark this answers is not in the document]
Comment (babamishra9) [reply zone-service/build.gradle]: hence there are only two string constants are required from sun.security.provider.X509Factory.BEGIN_CERT
sun.security.provider.X509Factory.END_CERT, changed as declared constants for the same and removed the import sun.security.provider.X509Factory.
I have excluded gradle compiler args from build.gradle file.

Done. Thanks!

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java (hunk 2 above)
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java]: Necessary, qualifiers and required parameters have been kept as part of constructer inject of service.

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java (hunk 4 above)
Comment (babamishra9) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneService.java]: Yes, its not required, i have removed it accordingly. Hence these are there for broker set the authorities as rest url calls. thanks!

**Comment by babamishra9** [comment]
No longer required
